docs/module-guidelines.md

```diff
@@ -0,0 +1,184 @@
+# Module authoring guidelines
+
+Modules should adhere to following guidelines.
+
+## Terraform Requirements
+
+The module source field must point to a single terraform module. We recommend
+the following structure:
+
+* `main.tf` file composing the terraform resources using provided variables.
+* `variables.tf` file defining the variables used.
+* (Optional) `outputs.tf` file defining any exported outputs used (if any).
+* (Optional) `modules/` sub-directory pointing to submodules needed to create the
+  top level module.
+
+## General Best Practices
+
+* Variables for environment-specific values (like `project_id`) should not be
+  given defaults. This forces the calling module to provide meaningful values.
+* Variables should only have zero-value defaults (like null or empty strings)
```

Comment:
nit: I strongly prefer null as an empty value to empty string. Perhaps just adding

s/(like null or empty strings)/, such as null (prefered) or empty strings,

Reply:
+1

Reply:
The value `null` is special and we are probably not thinking about it quite right. In particular, the `nullable=false` attribute on a module input variable will cause the module to use the default (not null) value if a user intentionally supplies null. This is perfect for module wrappers to respect the default values of the underlying module.

Easiest summary I could find:

https://blog.nuvotex.de/terraform-default-values-in-nested-modules/

So a default value of `null` in a wrapper is great, perfect, etc. In the module being wrapped, I think we need more careful consideration about what `null` conveys. If the variable is of type list and one iterates over it, _it's_ extremely nice to have a default empty list rather than null.

Additionally, I think a lot of our modules are being written with functional language approaches: e.g. `var.x != null` instead of `coalesce(var.x, default_value)`.

All that said, empty string isn't great. I suggest that, for now, we replace "string" with "list".